File: src/mk1/PythonPackage/src/tamaf/tickerbehavior.py
import time
from abc import abstractmethod
from .behavior import Behavior

class TickerBehavior(Behavior):
    def __init__(self, agent, tickPeriod: int):
        super().__init__(agent)
        self.period = tickPeriod / 1000.0
        self.lastTick = time.time()

    def Action(self):
        now = time.time()
        if now >= self.lastTick + self.period:
            self.OnTick()
            self.lastTick = now

    @abstractmethod
    def OnTick(self):
        pass
    
    def Done(self):
        return False

File: src/mk2/JavaPackage/src/main/java/tamaf/TickerBehavior.java
package tamaf;

public abstract class TickerBehavior extends Behavior {
    private long period;
    private long lastTick;

    public TickerBehavior(long tickPeriod) {
        super();
        this.period = tickPeriod;
        this.lastTick = System.currentTimeMillis();
    }

    public TickerBehavior(Agent agent, long tickPeriod) {
        super(agent);
        this.period = tickPeriod;
        this.lastTick = System.currentTimeMillis();
    }

    @Override
    public void action() {
        long now = System.currentTimeMillis();
        if (now >= lastTick + period) {
            onTick();
            lastTick = now;
        }
    }

    public abstract void onTick();

    @Override
    public boolean done() {
        return false;
    }
}
